### DISCUSSION TITLE
Global functions to programatically create nodes

### DISCUSSION THREAD

**Discussion by tslocke** [ueberdosis/tiptap#2090] [General Questions]
It would be convenient to expose the functions to create the various node types in my schema as module exports.

Once I have the schema, I can access these as, e.g. `schema.nodes.paragraph(...)` but with TipTap the schema is only available after the editor is created. I end up with the same code unpacking these from `schema.nodes` all over the place.

I was thinking there might be a way to access the ProseMirror NodeType from the TipTap Node subclass. Then I could do something like:

```
export const Paragraph = Node.create({...})

export const paragraph = (...args) => Paragraph.???.create(...args)
```

but it doesn't look like that's available. Is there a way to do something like this?

**Comment by philippkuehn** [comment]
I don't quite understand what you're trying to do yet but are you looking for [`extend`](https://tiptap.dev/guide/custom-extensions#extend-existing-extensions)?

```js
export const paragraph = (...args) => Paragraph.extend(...args)
```

**Reply by tslocke** [reply]
I want a bunch of global functions to create *instances* of the various node types in my schema. e.g. like ProseMirror's `create` method on NodeType. The first snippet in my post is wrong - I meant `schema.nodes.paragraph.create(...)` - sorry for the confusion!

**Reply by philippkuehn** [reply]
Ah ok. No, there is not such an API because the schema is not created within these extensions. 

Not sure if this helps but you can create a schema based on your extensions:

```js
import { getSchema, Node } from '@tiptap/core'

const Paragraph = Node.create({...})
const AnotherNode = Node.create({...})
// ...

const schema = getSchema([
  Paragraph,
  AnotherNode,
  // ...
])

const node = schema.nodes.paragraph.create(...)
```

**Reply by tslocke** [reply]
Yes! That's very helpful, combined with the trick you shared in #2077. I can now create a global schema from my TipTap node extensions, set it as the active TipTap schema using `onBeforeCreate`, and make the ProseMirror NodeTypes available as module exports:

```js
const {paragraph, heading, ...} = schema.nodes
export {paragraph, heading, ...}
```

These are very useful to have around, e.g. for testing if `node.type == heading` or for `heading.create(...)`

**Reply by philippkuehn** [reply]
Great! And within node and mark extensions you can always use `this.type` to get the ProseMirror node/mark type (after the schema is created of course)
